Version history for v4 APIs - Part 1: version history table and version details

JedrzejJanasiak opened the pull request:
## Issue

https://gravitee.atlassian.net/browse/APIM-3912 

## Description

This task contains changes required for releasing Version history for v4 APIS covering: 

 1. Version History screen with deployed API version 
![Zrzut ekranu 2024-02-19 140325](https://github.com/gravitee-io/gravitee-api-management/assets/155629548/85ebf3b8-4f65-4d2e-b888-fe45809d8e9f)

 2. API deployment preview with JSON Api definition
![Zrzut ekranu 2024-02-19 140346](https://github.com/gravitee-io/gravitee-api-management/assets/155629548/18f62627-3d87-4416-983d-c930dacc65d4)

 This task should cover listed below Acceptance Criteria:

- User can view a paginated list of the version history supporting at least 1,000 deployments.
- User can view the version history of their API.


<!-- Environment placeholder -->

🏗️ Your changes can be tested here and will be available soon:
      Console: [https://pr.team-apim.gravitee.dev/6652/console](https://pr.team-apim.gravitee.dev/6652/console)
      Portal: [https://pr.team-apim.gravitee.dev/6652/portal](https://pr.team-apim.gravitee.dev/6652/portal)
      Management-api: [https://pr.team-apim.gravitee.dev/6652/api/management](https://pr.team-apim.gravitee.dev/6652/api/management)
      Gateway v4: [https://pr.team-apim.gravitee.dev/6652](https://pr.team-apim.gravitee.dev/6652)
      Gateway v3: [https://pr.gateway-v3.team-apim.gravitee.dev/6652](https://pr.gateway-v3.team-apim.gravitee.dev/6652)

<!-- Environment placeholder end -->
<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-rprzttmhnp.chromatic.com)
<!-- Storybook placeholder end -->

Commit:
feat: migration of backend endpoint for fetching single event by its id

Commit:
feat(console): deployment history for v4 APIs

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployment-info/deployment-info.component.html:
```diff
@@ -0,0 +1,49 @@
+          <p><b>Label:</b> {{ apiDeploymentEvent.properties[LABEL_PROPERTY] }}</p>
+        </div>
+        <div class="card__codeblock">
+          <fieldset>
```
IMHO Should be this component from ui-particles in readonly
![image](https://github.com/gravitee-io/gravitee-api-management/assets/4974420/ab122637-da08-479a-8d88-647cdd2aacd3)
The figma models don't look right to me

JedrzejJanasiak replied:
done

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployment-info/deployment-info.component.scss:
```diff
@@ -0,0 +1,56 @@
+      width: 50%;
+
+      button {
+        position: absolute;
```
I'm not sure it's necessary here. absolute should really be avoided. and only used if necessary.

JedrzejJanasiak replied:
Yeah, you're right. It turned out that this part was unnecessary. I just removed it

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployment-info/deployment-info.component.scss:
```diff
@@ -0,0 +1,56 @@
+    justify-content: space-between;
+
+    &__title {
+      width: 50%;
```
The widths depending on the page in % are often "strange". What do you think of using flex ? (as it has inside `display: flex;`) ?

JedrzejJanasiak replied:
Same here, unnecessary and removed

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployment-info/deployment-info.component.spec.ts:
```diff
@@ -0,0 +1,66 @@
+
+  function expectApiEventRequest(response = fakeEvent()) {
+    const req = httpTestingController.expectOne(`${CONSTANTS_TESTING.env.v2BaseURL}/apis/${API_ID}/events/${API_VERSION_ID}`);
+    if (!req.cancelled) req.flush(response);
```
if you have canceled request, maybe the component is making several calls when it shouldn't be, and you need to correct it rather than hide this in the tests?

JedrzejJanasiak replied:
I suppose some mindless copy paste from my side. Fortunately, there was no redundant calls in this component so I just removed this `if (!req.cancelled)` part and it still passes tests

JedrzejJanasiak commented:
@ThibaudAV 
Updated views according to your suggestions: 

![Zrzut ekranu 2024-02-27 083710](https://github.com/gravitee-io/gravitee-api-management/assets/155629548/09e0fafc-4e9e-4b08-9705-d2012a77d46c)
![Zrzut ekranu 2024-02-27 084605](https://github.com/gravitee-io/gravitee-api-management/assets/155629548/87f89550-1628-4cb0-ac4b-8cc5d40cf073)

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/apis-routing.module.ts:
```diff
@@ -974,6 +976,30 @@ const apisRoutes: Routes = [
+        path: 'v4/history/:apiVersionId',
+        component: DeploymentInfoComponent,
+        data: {
+          apiPermissions: {
```
same ;)

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/apis-routing.module.ts:
```diff
@@ -974,6 +976,30 @@ const apisRoutes: Routes = [
+        data: {
+          apiPermissions: {
+            only: ['api-event-r'],
+          },
```
You'll have to rebase, but now this part is :
```suggestion
          permissions: {
            anyOf: ['api-event-r'],
          },
```

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/api-history-v4.component.scss:
```diff
@@ -0,0 +1,22 @@
+}
+
+.history {
+  padding: 24px;
```
i think ou can keep default `mat-card-content` padding. 
Later I'd like it to be defined by default for all the application. So that we don't have to define it page by page.

JedrzejJanasiak replied:
done

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/api-history-v4.component.spec.ts:
```diff
@@ -0,0 +1,141 @@
+
+  beforeEach(async () => {
+    TestBed.configureTestingModule({
+      imports: [NoopAnimationsModule, MatIconTestingModule, GioTestingModule, MatTableModule, MatCardModule, GioTableWrapperModule],
```
to avoid redefining all the component imports, we generally try to use the module (`ApiHistoryV4Module`) here and only have the imports linked to the test.

JedrzejJanasiak replied:
Good point :) I left only NoopAnimationsModule and two test modules

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/api-history-v4.module.ts:
```diff
@@ -0,0 +1,63 @@
+    GioTableWrapperModule,
+    MatTooltipModule,
+    MatPaginatorModule,
+    RouterLink,
```
needed ? 🤔

JedrzejJanasiak replied:
Indeed, I use it for routing to deployment info view as a button property: 
```
<button
   mat-icon-button
   aria-label="Button to show version"
   matTooltip="Show version"
   routerLink="{{ element.id }}"
   id="actionButton">
   <mat-icon svgIcon="gio:eye-empty"></mat-icon>
</button>
```

Unless there is other way to do it better?

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployment-info/deployment-info.component.scss:
```diff
@@ -0,0 +1,33 @@
+}
+
+.card {
+  padding: 24px;
```
same

JedrzejJanasiak replied:
done

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployments-table/deployments-table.component.scss:
```diff
@@ -0,0 +1,14 @@
+tr td:first-child,
```
i's better here to use class than direct html tag
you can use `.mat-column-version` or `.mat-column-action` it's more readable. 
more info : [here](https://v16.material.angular.io/cdk/table/overview#styling-columns)

JedrzejJanasiak replied:
Didn't know about that! Done

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployment-info/deployment-info.component.spec.ts:
```diff
@@ -0,0 +1,66 @@
+    httpTestingController.verify();
+  });
+
+  it('should create component with proper page title', async () => {
```
maybe you can also expect page data,  user, label  ?

JedrzejJanasiak replied:
Sure, no problem. Added to test

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/deployment-info/deployment-info.component.spec.ts:
```diff
@@ -0,0 +1,66 @@
+
+  beforeEach(async () => {
+    await TestBed.configureTestingModule({
+      imports: [GioTestingModule],
```
it's strange to not see MatCardModule, ... here. no ?  i think you also can directly import this module `ApiHistoryV4Module` and remove declarations here 🤔

JedrzejJanasiak replied:
Sure, good pint. Done

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/history-v4/api-history-v4.component.html:
```diff
@@ -0,0 +1,35 @@
+      </div>
+    </div>
+  </mat-card-content>
+  <ng-container *ngIf="apiEvents$ | async as apiDeployments">
```
I've noticed that you've used as many async pipes as possible, which is really good practice 💪

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/apis-routing.module.ts:
```diff
@@ -979,6 +981,30 @@ const apisRoutes: Routes = [
+        path: 'v4/history',
+        component: ApiHistoryV4Component,
+        data: {
+          apiPermissions: {
```
```suggestion
          permissions: {
```

JedrzejJanasiak replied:
done

ThibaudAV commented on gravitee-apim-console-webui/src/management/api/apis-routing.module.ts:
```diff
@@ -979,6 +981,30 @@ const apisRoutes: Routes = [
+        path: 'v4/history/:apiVersionId',
+        component: DeploymentInfoComponent,
+        data: {
+          apiPermissions: {
```
```suggestion
          permissions: {
```

JedrzejJanasiak replied:
done

ThibaudAV reviewed (approved):
- 2 last things
- And perhaps it remains that the state of if the list is empty display that it is empty in the table (some example with others table)

Code LGTM. but maybe one of your team can also review and test no ?

JedrzejJanasiak commented:
@ThibaudAV 

I added the empty table state the same way as the other components and added a test to cover this scenario. I also replaced the simple `<p>` tag when a table is loading with a dedicated gio-loading component. I hope that's fine with you.

ytvnr commented on gravitee-apim-console-webui/src/management/api/apis-routing.module.ts:
```diff
@@ -79,6 +79,8 @@ import { ApiPropertiesComponent } from './properties/properties/api-properties.c
 import { ApiRuntimeAlertsComponent } from './runtime-alerts';
 import { ApiDocumentationV4MetadataComponent } from './documentation-v4/documentation-metadata/api-documentation-v4-metadata.component';
+import { ApiHistoryV4Component } from './history-v4/api-history-v4.component';
+import { DeploymentInfoComponent } from './history-v4/deployment-info/deployment-info.component';
```
It is just a detail, but we often try to have consistent naming for component of module
no problem with the folder name, but I would have named the file `api-history-v4-deployment-info` (and rename the component in the same fashion)
@Okhelifi @ThibaudAV I would be curious to have your thoughts on this point

JedrzejJanasiak replied:
sure, I changed it as you proposed to keep it aligned with the rest of the component's names

ThibaudAV replied:
I like it too, but it creates components with names that are too long. So you have to find a compromise between a clear name with "parent name" and "component name"
in that case it's not too long :p

ytvnr commented on gravitee-apim-console-webui/src/management/api/apis.module.ts:
```diff
@@ -98,6 +99,7 @@ import { AlertsModule } from '../../components/alerts/alerts.module';
     DocumentationModule,
     GioPolicyStudioRoutingModule,
     SpecificJsonSchemaTypeModule,
+    ApiHistoryV4Module,
```
Can you put it up in the list, I think we tried to keep the alphabetical order 🤔

JedrzejJanasiak replied:
no problemo, done

ytvnr commented on gravitee-apim-console-webui/src/management/api/history-v4/api-history-v4.component.html:
```diff
@@ -0,0 +1,38 @@
+    </div>
+  </mat-card-content>
+  <ng-container *ngIf="apiEvents$ | async as apiDeployments; else loading">
+    <app-deployments-table
```
change the selector name to not have `app` prefix, maybe choose `api-deployments-table`

JedrzejJanasiak replied:
done, changed to `api-deployments-table` as suggested

ytvnr commented on gravitee-apim-console-webui/src/management/api/history-v4/api-history-v4.component.ts:
```diff
@@ -0,0 +1,53 @@
+export class ApiHistoryV4Component {
+  private apiId = this.activatedRoute.snapshot.params.apiId;
+
+  protected search = new BehaviorSubject<SearchApiEventParam>(INITIAL_SEARCH_PARAM);
```
just about naming, but I would have chosen something like `filters$`
`search` makes me feel like the search will be done, when it is just a way to observe the filter to use for the request

JedrzejJanasiak replied:
good point, done

ytvnr reviewed (commented):
looks good to me, good job!
some minor remarks but we can discuss it
